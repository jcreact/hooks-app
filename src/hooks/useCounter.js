import { useState } from 'react';

/**
 * Gestiona el estado de un contador.
 * @customHook
 * 
 * @param {number} [value = 0] Es el valor inicial del contador. 
 * @returns {{
 *   state: number, 
 *   increment: (factor?: number) => void, 
 *   decrement: (factor?: number) => void
 * }}
 *   Hook para el manejo de un contador.
 */
export const useCounter = (value = 0) => {

    const [state, setState] = useState(value);

    const increment = (factor = 1) => {
        setState(s => s + factor);
    };

    const decrement = (factor = 1) => {
        setState(s => s - factor);
    };

    return {
        state,
        increment,
        decrement
    };

};