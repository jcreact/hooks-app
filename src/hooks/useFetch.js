import { useEffect, useState } from 'react';

export const useFetch = (url) => {

    const [state, setState] = useState({ data: null, error: null, loading: true });

    const reset = () => setState({
        data: null,
        error: null,
        loading: true
    });

    useEffect(() => {
        reset();
        fetch(url)
            .then(resp => resp.json())
            .then(data =>
                setState({
                    loading: false,
                    error: null,
                    data
                })
            );

    }, [url]);

    return state;

};
