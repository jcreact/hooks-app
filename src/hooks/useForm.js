import { useState } from 'react';


export const useForm = (formValues = {}) => {
    const [values, setValues] = useState(formValues);

    /**
     * Maneja el evento onChange de los valores del formulario.
     *
     * @param {React.ChangeEvent<HTMLInputElement>} evt
     */
    const handleInputChange = ({ target }) => {
        setValues({ ...values, [target.name]: target.value });
    };

    const reset = () => {
        setValues(formValues);
    };

    return [values, handleInputChange, reset];

};