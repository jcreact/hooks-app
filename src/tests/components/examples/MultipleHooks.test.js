import React from 'react';
import { shallow } from 'enzyme';

import { MultipleHooks } from '../../../components/examples/MultipleHooks.jsx';

import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

// Mock Hook
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe(`Pruebas en el componente 'MultipleHooks.jsx'.`, () => {

    const increment = jest.fn();

    beforeEach(() => {
        jest.resetAllMocks();
        useCounter.mockReturnValue({
            counter: 1,
            increment
        });
    });

    test(`Debe renderizar el componente MultipleHooks correctamente.`, () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const component = shallow(<MultipleHooks />);
        expect(component).toMatchSnapshot();
        expect(component.find('Spinner').exists()).toBeTruthy();
        expect(component.find('button').prop('disabled')).toBeTruthy();

    });

    test(`Debe renderizar una cita.`, () => {

        const author = 'José Palma';
        const quote = '!Es todo un show!';
        const series = 'La Vida';

        // Simulando la información del hook.
        useFetch.mockReturnValue({
            data: [{
                author,
                quote,
                series
            }],
            loading: false,
            error: null
        });

        const component = shallow(<MultipleHooks />);
        expect(component).toMatchSnapshot();
        expect(component.find('Spinner').exists()).toBeFalsy();

        expect(component.find('blockquote').exists()).toBeTruthy();
        expect(component.find('blockquote').find('p').text()).toContain(quote);

        expect(component.find('figcaption').exists()).toBeTruthy();
        expect(component.find('figcaption').text()).toContain(author);
        expect(component.find('figcaption').text()).toContain(series);

        expect(component.find('button').prop('disabled')).toBeFalsy();
        component.find('button').simulate('click');
        expect(increment).toBeCalled();

    });

});
