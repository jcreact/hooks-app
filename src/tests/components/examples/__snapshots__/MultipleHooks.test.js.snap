// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente 'MultipleHooks.jsx'. Debe renderizar el componente MultipleHooks correctamente. 1`] = `
<Fragment>
  <h1
    className="mt-3"
  >
    BreakingBad Quotes
  </h1>
  <hr />
  <Spinner />
  <button
    className="btn btn-outline-primary"
    disabled={true}
    onClick={[MockFunction]}
  >
    <i
      className="bi bi-caret-right-fill"
    />
     Siguiente
  </button>
</Fragment>
`;

exports[`Pruebas en el componente 'MultipleHooks.jsx'. Debe renderizar una cita. 1`] = `
<Fragment>
  <h1
    className="mt-3"
  >
    BreakingBad Quotes
  </h1>
  <hr />
  <figure
    className="text-end border border-1 border-success rounded-3 shadow"
  >
    <blockquote
      className="blockquote me-3 mt-2"
    >
      <p>
        !Es todo un show!
      </p>
    </blockquote>
    <figcaption
      className="blockquote-footer me-3"
    >
      Jos√© Palma
       
      <cite>
        "
        La Vida
        "
      </cite>
    </figcaption>
  </figure>
  <button
    className="btn btn-outline-primary"
    disabled={false}
    onClick={[MockFunction]}
  >
    <i
      className="bi bi-caret-right-fill"
    />
     Siguiente
  </button>
</Fragment>
`;
